Sources/Utility/OutputByteStream.swift

```diff
@@ -0,0 +1,359 @@
+ Licensed under Apache License v2.0 with Runtime Library Exception
+
+ See http://swift.org/LICENSE.txt for license information
+ See http://swift.org/CONTRIBUTORS.txt for Swift project authors
+*/
+
+/// Convert an integer in 0..<16 to its hexadecimal ASCII character.
+private func hexdigit(_ value: UInt8) -> UInt8 {
+    return value < 10 ? (0x30 + value) : (0x41 + value - 10)
+}
+
+/// Describes a type which can be written to a byte stream.
+public protocol ByteStreamable {
+    func writeTo(_ stream: OutputByteStream)
+}
+
+/// An output byte stream.
+///
+/// This class is designed to be able to support efficient streaming to
+/// different output destinations, e.g., a file or an in memory buffer. This is
+/// loosely modeled on LLVM's llvm::raw_ostream class.
+///
+/// The stream is generally used in conjunction with the custom streaming
+/// operator '<<<'. For example:
+///
+///   let stream = OutputByteStream()
+///   stream <<< "Hello, world!"
+///
+/// would write the UTF8 encoding of "Hello, world!" to the stream.
+///
+/// The stream accepts a number of custom formatting operators which are defined
+/// in the `Format` struct (used for namespacing purposes). For example:
+/// 
+///   let items = ["hello", "world"]
+///   stream <<< Format.asSeparatedList(items, separator: " ")
+///
+/// would write each item in the list to the stream, separating them with a
+/// space.
+public class OutputByteStream: OutputStream {
+    /// The data buffer.
+    private var buffer: [UInt8]
+    
+    public init() {
+        self.buffer = []
+    }
+
+    // MARK: Data Access API
+
+    /// The current offset within the output stream.
+    public var position: Int {
+        return buffer.count
+    }
+
+    /// The contents of the output stream.
+    ///
+    /// This method implicitly flushes the stream.
+    public var bytes: ByteString {
+        flush()
+        return ByteString(self.buffer)
+    }
+    
+    // MARK: Data Output API
+
+    public func flush() {
+        // Do nothing.
+    }
+
+    /// Write an individual byte to the buffer.
+    public func write(_ byte: UInt8) {
+        buffer.append(byte)
+    }
+    
+    /// Write a sequence of bytes to the buffer.
+    public func write(_ bytes: [UInt8]) {
+        buffer += bytes
+    }
+    
+    /// Write a sequence of bytes to the buffer.
+    public func write(_ bytes: ArraySlice<UInt8>) {
+        buffer += bytes
+    }
+    
+    /// Write a sequence of bytes to the buffer.
+    public func write<S: Sequence where S.Iterator.Element == UInt8>(_ sequence: S) {
+        buffer += sequence
+    }
+
+    /// Write a string to the buffer (as UTF8).
+    public func write(_ string: String) {
+        // Fast path for contiguous strings. For some reason Swift itself
+        // doesn't implement this optimization: <rdar://problem/24100375> Missing fast path for [UInt8] += String.UTF8View
+        let stringPtrStart = string._contiguousUTF8
```

Comment:
You want to add `defer { _fixLifetime(string) }` here, or the optimizer is allowed to shorten the lifetime of the string.